crates/orchestrator/src/jobs/state_update_job/kzg.rs

```diff
@@ -0,0 +1,124 @@
+use crate::jobs::state_update_job::CURRENT_PATH;
+use alloy::eips::eip4844::BYTES_PER_BLOB;
+use c_kzg::{Blob, Bytes32, KzgCommitment, KzgProof, KzgSettings};
+
+/// Build KZG proof for a given block
+pub async fn build_kzg_proof(
+    block_number: u64,
+    fetch_from_tests: Option<bool>,
+) -> color_eyre::Result<(Vec<u8>, KzgProof)> {
+    let blob_data = fetch_blob_data_for_block(block_number, fetch_from_tests).await?;
+    let mut fixed_size_blob: [u8; BYTES_PER_BLOB] = [0; BYTES_PER_BLOB];
+    fixed_size_blob.copy_from_slice(blob_data.as_slice());
+
+    let x_0_value = fetch_x_0_value_from_os_output(block_number, fetch_from_tests).await?;
+
+    // trusted setup ceremony
+    let trusted_setup_path = CURRENT_PATH.join("src/jobs/state_update_job/trusted_setup.txt");
+    let trusted_setup =
+        KzgSettings::load_trusted_setup_file(trusted_setup_path.as_path()).expect("Error loading trusted setup file");
+
+    let blob = Blob::new(fixed_size_blob);
+    let commitment = KzgCommitment::blob_to_kzg_commitment(&blob, &trusted_setup)?;
+    let (kzg_proof, y_0_value) = KzgProof::compute_kzg_proof(&blob, &x_0_value, &trusted_setup)?;
+
+    // Verifying the proof for double check
+    let eval = KzgProof::verify_kzg_proof(
+        &commitment.to_bytes(),
+        &x_0_value,
+        &y_0_value,
+        &kzg_proof.to_bytes(),
+        &trusted_setup,
+    )?;
+    assert!(eval);
+
+    Ok((blob_data, kzg_proof))
+}
+
+/// Fetching the blob data (stored in s3 during DA job) for a particular block
+pub async fn fetch_blob_data_for_block(
+    block_number: u64,
+    fetch_from_tests: Option<bool>,
+) -> color_eyre::Result<Vec<u8>> {
+    let fetch_from_tests = fetch_from_tests.unwrap_or(true);
+    let blob_data: Vec<u8> = match fetch_from_tests {
+        true => {
+            let blob_data_path =
+                CURRENT_PATH.join(format!("src/jobs/state_update_job/test_data/{}/blob_data.txt", block_number));
+            let data = std::fs::read_to_string(blob_data_path).expect("Failed to read the blob data txt file");
+            hex_string_to_u8_vec(&data).unwrap()
+        }
+        false => {
+            let s3_client = AWSS3::new(AWSS3Config::new_from_env()).await;
+            let key = block_number.to_string() + "/blob_data.txt";
+            let blob_data = s3_client.get_data(&key).await?;
+            blob_data.to_vec()
+        }
+    };
+
+    Ok(blob_data)
+}
+
+pub async fn fetch_x_0_value_from_os_output(
+    block_number: u64,
+    fetch_from_tests: Option<bool>,
+) -> color_eyre::Result<Bytes32> {
+    let fetch_from_tests = fetch_from_tests.unwrap_or(true);
+    let x_0 = match fetch_from_tests {
+        true => {
+            let x_0_path = CURRENT_PATH.join(format!("src/jobs/state_update_job/test_data/{}/x_0.txt", block_number));
+            let data = std::fs::read_to_string(x_0_path)?;
+            Bytes32::from_hex(&data).unwrap()
+        }
+        false => unimplemented!(),
```

Comment:
do we know how to fetch this now given we know the SNOS output = calldata on Eth

Reply:
so in our case when SNOS run is complete then we will store the output and we store x_0 point into another file. wdyt ?